### PR TITLE
Add time-frequency resolved DICS beamformer

### PR DESCRIPTION
**Pull request by alexrockhill**
The beamformer methods part of https://github.com/mne-tools/mne-python/pull/10920.

### CODE REVIEW

**Comment by alexrockhill** [comment]
So the one thing I'm not sure about is that the DICS beamformer is mostly used with `apply_dics_csd` which is a spectral, time-frequency estimate but there exists `apply_dics_epochs` although it says it has not been tested thoroughly with a warning. I don't see anywhere in the code where the epochs are transformed to a time-frequency estimate; the beamformer filter is applied directly to the time-series data as far as I can tell. In this PR, it is applied to time-frequency resolved epochs following the CSD approach. Seems like a potential consistency issue but maybe I'm just missing something.

**Comment by alexrockhill** [comment]
It looks like the test failures are unrelated and are due to a new `pyvista` release.

**Comment by alexrockhill** [comment]
I added a test to compare with `apply_dics_csd` which gives me a lot more confidence that it is implemented correctly since it is widely used/validated.

**Comment by larsoner** [comment]
@wmvanvliet can you look?

**Comment by britta-wstnr** [comment]
Re application on epochs / evoked: this is what @wmvanvliet and I have said (in one of the other PRs )was old code from before we touched the code base. I would not recommend this approach, and as I understood @wmvanvliet thinks the same?

**Comment by agramfort** [comment]
yes to me it's not standard what is done here. As I said elsewhere I would suggest to focus on MNE based methods and the GUI that can use the MNE outputs

**Comment by wmvanvliet** [comment]
This PR concerns applying DICS beamformer weights to a (complex-valued) EpochsTFR object. To my understanding, that is the approach advocated in [Laaksonen, Kujala & Salmelin (2008)](https://doi.org/10.1016/j.neuroimage.2008.04.175). I agree with keeping the tutorial focused on MNE based methods, but having this `apply_dics_tfr_epochs` function around in its own PR is a good idea. This is something we would want to add in terms of DICS functionality.

File: mne/beamformer/tests/test_dics.py
Comment (britta-wstnr) [inline mne/beamformer/tests/test_dics.py:437]:
```suggestion
    # Also test here that no warnings are thrown - implemented to check whether
```

File: mne/beamformer/tests/test_dics.py
Comment (britta-wstnr) [inline mne/beamformer/tests/test_dics.py:581]:
```suggestion
    """Test DICS applied to time-frequency objects."""
```

File: mne/beamformer/tests/test_dics.py
Comment (britta-wstnr) [inline mne/beamformer/tests/test_dics.py:598]: This whole test function could use some more comments on what you are testing where (i.e., real data, complex data etc). Will make it easier to trouble shoot in the future if necessary.

File: mne/beamformer/tests/test_dics.py
Comment (britta-wstnr) [inline mne/beamformer/tests/test_dics.py:631]: Why compatibility with Matlab?

File: mne/beamformer/tests/test_dics.py
Comment (britta-wstnr) [inline mne/beamformer/tests/test_dics.py:622]: This does not directly concern the DICS TFR, does it?

File: mne/beamformer/tests/test_dics.py
Comment (britta-wstnr) [reply mne/beamformer/tests/test_dics.py:631]: I acutally misunderstood your intentions here first - again, add comments further up to explain what this test will do

File: mne/beamformer/tests/test_dics.py
Comment (britta-wstnr) [reply mne/beamformer/tests/test_dics.py:622]: I acutally misunderstood your intentions here first - again, add comments further up to explain what this test will do

Comment (britta-wstnr) [reply mne/beamformer/tests/test_dics.py:622]: (EDIT: I had initially replied the last comment in the wrong thread.)

**Comment by wmvanvliet** [comment]
Regarding whether applying DICS weights to a TFR object is a good idea, we asked around at the BioMag conference. As far as I could tell, Jan Kujala, Robert Oostenveld, Britta Westner, and I are all +1 on this idea. I'm not aware of any publications that actually do this though (Jan Kujala has one paper in the pipeline in which this technique is used, but it's not published yet).

File: mne/beamformer/tests/test_dics.py
Comment (alexrockhill) [reply mne/beamformer/tests/test_dics.py:622]: No, it does not directly concern the content of this (applying to a non-csd time-frequency object), it's just since the frequencies are taken as averages of high and low bounds (or in this case only one bound) I thought it was prudent to check.

File: mne/beamformer/tests/test_dics.py
Comment (alexrockhill) [reply mne/beamformer/tests/test_dics.py:631]: This is copied verbatim from the csd code so that the same process and normalizations are followed. That's a good question which I might guess that the answer is that the original MNE implementation was a translation from Matlab.

**Comment by agramfort** [comment]
honestly I would not push too much here. Let's focus our time and attention on the GUI code.

my 2c

**Comment by alexrockhill** [comment]
> honestly I would not push too much here. Let's focus our time and attention on the GUI code.
> 
> my 2c

Sounds good, I'll leave off here.

File: mne/beamformer/tests/test_dics.py
Comment (wmvanvliet) [reply mne/beamformer/tests/test_dics.py:634]: Possibly to more closely reproduce what Fieldtrip does?

**Comment by wmvanvliet** [comment]
I've cleaned up the code for the test a bit, and added more comments. If you check the 'Allow edits from maintainers' box, I can push directly to this PR. For now, the code is here: https://gist.github.com/wmvanvliet/294d418618bf53847bbbd1d132cd7c9d

**Comment by wmvanvliet** [comment]
Silly me, I was able to push all along.

**Comment by alexrockhill** [comment]
Oh well I guess there is one main holdup for merging which is that the topo plot appears empty here: https://output.circle-artifacts.com/output/job/82ee181a-3845-4871-b9ee-b5bb913b5b44/artifacts/0/dev/auto_tutorials/inverse/55_tfr_stc.html

When I pass one channels in picks it shows up just fine but if I pass ten channels like every third or so is blank. It's a very odd bug, not sure what's going on but it appears to have broken recently as it was working just fine earlier in this PR... I don't think it has to do with converting complex phase-amplitude to amplitude since it does work with a subselection of picks. Really not sure what's going on there...

**Review by wmvanvliet — APPROVED** [review]

**Comment by wmvanvliet** [comment]
Wait, one last thing. There should be a small example demonstrating how to apply this function. Not a complete tutorial, but more of a minimum working example type of thing that users can copy and paste.

**Comment by alexrockhill** [comment]
> Wait, one last thing. There should be a small example demonstrating how to apply this function. Not a complete tutorial, but more of a minimum working example type of thing that users can copy and paste.

Can we merge this and then add the example in https://github.com/mne-tools/mne-python/pull/10920? I think it would be nice to have copy-paste-able code to compare both methods.

File: mne/beamformer/tests/test_dics.py
Comment (britta-wstnr) [reply mne/beamformer/tests/test_dics.py:437]: Can you commit this?

**Comment by wmvanvliet** [comment]
> Can we merge this and then add the example in https://github.com/mne-tools/mne-python/pull/10920?

Give me a minute. I'll write up a quick example.

**Comment by wmvanvliet** [comment]
Ok, the example seems to produce something sensible. However, an issue pops up. If you feed in a complex-valued TFR to the DICS filter, you will get a complex-valued `SourceEstime` object. Does @alexrockhill's new plotting code deal with this somehow? In the example I now do `stc.data = np.abs(stc.data)` as a quick hack.

**Comment by wmvanvliet** [comment]
This of course breaks when `pick_ori='vector'`...

**Comment by wmvanvliet** [comment]
Link to example: https://output.circle-artifacts.com/output/job/b680c779-e20c-4dae-9b0e-e40ad44facf1/artifacts/0/dev/auto_examples/inverse/dics_epochs.html

**Comment by alexrockhill** [comment]
Yes the plotting code allows toggling between average power, single trial power and ITC. Thanks for making an example!

**Comment by alexrockhill** [comment]
Also in the GUI if you have vector orientation it gives a warning if you try and do ITC because you don't really have a sensible orientation to compare across, it's better to do max power orientation or something to compare phase alignments. The x y and z orientations are going to be suboptimal representations of the oscillation in all likelihood, so you want max power I think.

File: examples/inverse/dics_epochs.py
Comment (alexrockhill) [inline examples/inverse/dics_epochs.py:82]: I think this is actually amplitude, if I'm not mistaken power would be:

```suggestion
# At this point, the data is still complex. Convert to power.
stc.data = (stc.data * stc.data.conj()).real
```

**Comment by alexrockhill** [comment]
Just one minor comment on @wmvanvliet 's addition of the example. Looks great, thanks for doing that! Very nice on the somato dataset where you can clearly see S1/M1 activation.

File: examples/inverse/dics_epochs.py
Comment (wmvanvliet) [reply examples/inverse/dics_epochs.py:82]: Agreed!

**Comment by wmvanvliet** [comment]
The example has allowed me to play more with the behavior of the DICS applied in this manner. I can verify that the `EpochsTFR` indeed must be complex-valued, otherwise you get garbage. However, it's not clear to @britta-wstnr and myself why it would matter so much to drop phase information. We even have `real_filter=True` enabled to disregard phase while building the DICS filter (setting `real_filter=False` gives good output too).

And there is still the issue of `pick_ori='vector'`. I think we should just disable that for now, because orientations cannot be sensible if we have to do `np.abs(stc.data)`.

**Comment by wmvanvliet** [comment]
I'm just writing down these things so we don't forget or accidentally merge this before we are sure about these things. We should focus on the GUI for now and not get too deep into this ;)

**Comment by larsoner** [comment]
@wmvanvliet we are close to 1.2 release, should we bump the milestone on this or is this a blocker?

**Comment by wmvanvliet** [comment]
Bump it. We're going to be tempted to merge this too hastily if we don't.

**Comment by alexrockhill** [comment]
I looked over the tutorial and, interestingly, the first epoch had much stronger activation than others so I plotted several epochs and several frequencies to show a bit more of the data.

**Comment by alexrockhill** [comment]
Ok to merge? I just looked over the example again and I think it looks good https://output.circle-artifacts.com/output/job/89cee0b1-b937-4bb3-9386-f76f9a0ecac5/artifacts/0/dev/auto_examples/inverse/dics_epochs.html#sphx-glr-auto-examples-inverse-dics-epochs-py. And, it shows some interesting inter-trial differences which might be helpful to understand what's going on in the dataset.

**Comment by larsoner** [comment]
@wmvanvliet can you look again when you get a chance?

**Comment by wmvanvliet** [comment]
So, what happens when I set `pick_ori='vector'`?

**Comment by alexrockhill** [comment]
> So, what happens when I set `pick_ori='vector'`?

Right now, the phase-amplitude of each x, y and z dimension is computed independently. I like that because I think seeing the direction of the amplitude is interesting, however this does cause issues with the phase being different at the same location. Perhaps it would be better to assign the `max-power` phase to all three dimensions? That seems like almost a good idea to me; that way you still have an orientation but you also can interpret the phase. The only issue is that it is going to effect the power calculations. Just thinking out loud here but perhaps we should let the user take the phase from the `max-power` computation if they are interested in the phase (we could provide an example of this but it would involve something like a phase-amplitude coupling analysis that is a pretty complicated use-case for an MNE example).

I think it sounds reasonable to take the phase from `max-power` and the amplitude from `vector` for these kinds of analysis or at least we should make it a possibility for users, what do you think?

**Comment by alexrockhill** [comment]
Hey, looks pretty cool when you plot the phase:

https://user-images.githubusercontent.com/13473576/196008102-be2ee678-70a2-4dda-8019-5dd4b638a297.mp4

I changed the other plots to the vector solution which looks pretty cool as well. I wouldn't use phase from it though since it would be 3D. Using the `max-power` phase seems to match pretty closely with what you would expect to see.

image

**Comment by alexrockhill** [comment]
Well I think this looks exceedingly cool https://output.circle-artifacts.com/output/job/77c91aff-ba58-4d71-ac93-4514777b2eb7/artifacts/0/dev/auto_examples/inverse/dics_epochs.html#sphx-glr-auto-examples-inverse-dics-epochs-py.

**Comment by agramfort** [comment]
@alexrockhill this makes indeed cool pictures !

now I am a bit worried about the message behind the single trial estimates? The data you use a raw without any artefact rejection so you may be interpreting or show noise. Can you check this? Saying that "activation varies quite a bit trial-to-trial" suggest it's real I would believe it just that single trials are simply too noisy for most trials. What do you think?

**Comment by alexrockhill** [comment]
> @alexrockhill this makes indeed cool pictures !
> 
> now I am a bit worried about the message behind the single trial estimates? The data you use a raw without any artefact rejection so you may be interpreting or show noise. Can you check this? Saying that "activation varies quite a bit trial-to-trial" suggest it's real I would believe it just that single trials are simply too noisy for most trials. What do you think?

There might be noisy trials but the activation varying trial-to-trial is referring to the activity at M1/S1 which isn't present on every trial which is interesting. It could be that the noise is corrupting the source estimate so much that it overpowers the M1/S1 activation though so we could try and fix that. I think this example looks okay and I would guess that the activation really isn't present but I can look into preprocessing the data a bit.

**Comment by agramfort** [comment]
>
> There might be noisy trials but the activation varying trial-to-trial is
> referring to the activity at M1/S1 which isn't present on every trial which
> is interesting. It could be that the noise is corrupting the source
> estimate so much that it overpowers the M1/S1 activation though so we could
> try and fix that.
>

at least that's my first intuition. As if it is not and we are sure about
this, then it's really intriguing.


> I think this example looks okay and I would guess that the activation
> really isn't present but I can look into preprocessing the data a bit.Message
> ID: ***@***.***>
>

try at least to use some reject for blinks, and even SSP correction just to
see how it affects the results.

**Comment by wmvanvliet** [comment]
> Right now, the phase-amplitude of each x, y and z dimension is computed independently. I like that because I think seeing the direction of the amplitude is interesting

Except you are not plotting the direction of the amplitude as the example is now. You are plotting this:
```
stc.data = (stc.data * np.conj(stc.data)).real
```
Meaning that the XYZ directions are all squared. Plotting the actual direction is way harder and probably involves computing the largest eigenvector of the source CSD at each timepoint. This is why, for now, I recommend disabling `apply_dics_tfr_epochs` for DICS beamformers that have been computed with `pick_ori='vector'`.

There are too many unknowns here. The reason we are so conservative in allowing cool but unproven stuff is that MNE-Python is relied on by researchers all over the world to produce analyses that are correct. If we mess things up, they will have to print a correction or even retraction of their paper if they actually used this. See this article: https://www.wired.co.uk/article/fmri-bug-brain-scans-results

The method employed by `apply_dics_tfr_epochs` is edgy enough by itself (remember, no actual publications have been published using this exact method!) without attempting novel visualization strategies too.

**Comment by alexrockhill** [comment]
That's a good point @wmvanvliet, maybe we should just go back to how it was before in the tutorial, the phase is kind of nice though.

I tried adding SSP and rejection thresholds and it still seems that there is not activation on every trial and that maybe even a minority of trials drive the effect. That wasn't what I was expecting...

**Comment by alexrockhill** [comment]
Just thinking about it, what if you set the power to be negative when the phase was between `pi` and `2 pi`? That seems so simple probably incredibly wrong but I'll just check what it looks like. Is there a reason that's a bad idea?

**Comment by alexrockhill** [comment]
Wow, okay I'm watching some of the movies of the vector solutions playing and they seem to be very interpretable on every trial, it's really quite awesome; you can see when they move a bit in the baseline period and M1 becomes active, you can see when they see the stimulus and V1 because active and then you can see when S1 becomes active sometime after the stimulus but the timing is really variable which is interesting. I think it's less an issue of activation on every trial; there's a lot of that, and more an issue of timing. I'll add a movie to the example and push.

**Comment by alexrockhill** [comment]
Looks biologically plausible to me, although it's hard to know if the vectors are meaningful https://user-images.githubusercontent.com/13473576/196270246-3e500886-793d-418e-ae82-e7643e6ea647.mp4

**Comment by larsoner** [comment]
>> Meaning that the XYZ directions are all squared... . This is why, for now, I recommend disabling apply_dics_tfr_epochs for DICS beamformers that have been computed with pick_ori='vector'.

> Looks biologically plausible to me, although it's hard to know if the vectors are meaningful

I agree with @wmvanvliet above where he implicitly says they are not, and this should not be allowed. To reiterate/rephrase, they will be constrained to the right-anterior-superior octant in 3D space because of the vector norm. You only get a sense of "how much in the X, Y, and Z" but it's not represented all that well, and it's misleading to users the way it's presented now.

> Just thinking about it, what if you set the power to be negative when the phase was between pi and 2 pi

I highly doubt this is correct because it can cause a huge jump / discontinuity if the phase shifts for example between 179 deg and 181 deg, which isn't plausible. Rather than trying to figure this out -- or even implement @wmvanvliet's probably correct solution using SVD -- let's save this stuff for a future PR and prevent people from doing/interpreting things in a dangerous way.

**Comment by alexrockhill** [comment]
> I highly doubt this is correct because it can cause a huge jump / discontinuity if the phase shifts for example between 179 deg and 181 deg, which isn't plausible.

Good point, I'll just go back to the tutorial as we had it before.

**Comment by alexrockhill** [comment]
That looks pretty good, seems like the movies are getting cut off at some point before they're supposed to though, I can open an issue https://output.circle-artifacts.com/output/job/d7454403-c62e-454a-8df1-365910c04437/artifacts/0/dev/auto_examples/inverse/dics_epochs.html#sphx-glr-auto-examples-inverse-dics-epochs-py.

I'm not going to keep bugging people on this PR with the attempts to make sense of the vector solution but I thought about it more and, compared to the real-valued solution which can be consistently positive or negative, the complex time-frequency valued solution only really makes sense constantly changing direction from negative to positive and back again so really viewing a static image isn't going to be very helpful. I'm going to explore it a bit more (outside of this PR) because I think it's interesting but you might be able to get good information out of taking `np.sin` of the phase and multiplying that by the power. That would solve the discontinuity issue @larsoner brought up. Obviously a full replication of something like traveling wave results would be required before it went into MNE but just thought I would mention it in case anyone had thoughts before checking it out a bit (probably not going to do a paper, just curious).

**Comment by alexrockhill** [comment]
https://user-images.githubusercontent.com/13473576/196536695-cb0546e7-f7af-4e48-b852-24f0afe65e51.mp4

Looks really interesting, again not for this PR, but I think there's some information there!

**Review by britta-wstnr — COMMENTED** [review]
Let me chime in here again.
Like @agramfort, I am not convinced about your single trial estimates. There are several reasons why this does not make much sense to me.

1. You plot them averaged across time (in your first plot where you show epochs) - the signal however is _induced_, such that we expect a time variability from trial to trial.
2. We can also expect some variability in peak frequency based on beta burst literature, which you are not capturing.
3. Most importantly: in single trials, your baseline will be very noisy and volatile. By performing baseline correction, you will carry that noise right into your signal of interest.

FWIW, I played around with the signal on sensor level tonight. If I do a dirty trick, which is **not recommended** to be used other than as a dirty hack, namely using the _averaged_ baseline for baseline correction of single epochs, I can recover the beta rebound. It has - as expected - a variability in strength, in peak frequency, and timing across trials. Some trials have clear artifacts in them (I looked at the first 10 and last 5 of the full dataset). 
What I would suggest is plot the _averaged_ TFR here. That is new for DICS beamformers on MNE-Python as well, and surely way more stable. It will not need any dangerous tricks such as averaging the baseline (which leads to a difference in magnitude, logically, a thing we should probably not promote to users who might go on to do stats etc on the output).

As for the rest, I agree with @larsoner and @wmvanvliet, that those phase-related solutions are not quite biophysically plausible.
Out of curiosity, in one of your comments you talk about V1 activity - I could not find any information about visual stimulation in this data set? All I were able to dig up is related to "pure" median nerve stimulation.

**Comment by alexrockhill** [comment]
> What I would suggest is plot the averaged TFR here.

Sounds reasonable, we can change it to that.

> We can also expect some variability in peak frequency based on beta burst literature, which you are not capturing.

We can also add a few more frequencies.

> Most importantly: in single trials, your baseline will be very noisy and volatile. By performing baseline correction, you will carry that noise right into your signal of interest.

Good point, I think, in the future, it would be interesting to see what kind of information can be gleaned at the single-trial level but that can be for a future PR or maybe a paper instead because as you suggest, there are some preprocessing decisions that need to be made.

> As for the rest, I agree with @larsoner and @wmvanvliet, that those phase-related solutions are not quite biophysically plausible.

Did you think that for the video above? It's definitely not going in this PR but I'm curious if it contains usable information. The `sin` version definitely looks a lot better than the positive/negative at least but I'd be very interested to know if there was something wrong with doing it that way.

> Out of curiosity, in one of your comments you talk about V1 activity - I could not find any information about visual stimulation in this data set? All I were able to dig up is related to "pure" median nerve stimulation.

I wish I knew more about how this dataset was collected. I was inferring that from the occipital activity right before time zero in the video above (and in the non-vector solution) but now that I write that, it's before time zero so I'm not sure if that makes sense as anything other than they were looking at something not time-locked to the trial that evoked some visual-related brain activity.

**Comment by alexrockhill** [comment]
Okay, this looks very reasonable and well inside what would be consider too experimental. I think pretty close to merging.

**Comment by alexrockhill** [comment]
Ok to merge?

Looks good to me https://output.circle-artifacts.com/output/job/b3ea7b04-13a1-4ff9-94d9-5a53163bfd6c/artifacts/0/dev/auto_examples/inverse/dics_epochs.html#sphx-glr-auto-examples-inverse-dics-epochs-py

**Review by agramfort — APPROVED** [review]

**Comment by agramfort** [comment]
thanks @alexrockhill 🙏

**Pull request merged**
